### ISSUE TITLE
More compact syntax for re-exported objects

### ISSUE THREAD

**Issue by yihui** [klutometis/roxygen#409]
I think this looks a little cumbersome (when we need to re-export multiple objects from multiple packages):

```r
#' @importFrom htmlwidgets JS saveWidget
#' @export
htmlwidgets::JS
#' @export
htmlwidgets::saveWidget
#' @importFrom magrittr %>%
#' @export
magrittr::`%>%`
```

It will be nicer if we can just

```r
#' @importFrom htmlwidgets JS saveWidget
#' @exportFrom htmlwidgets JS saveWidget
#' @importFrom magrittr %>%
#' @exportFrom magrittr %>%
```

Even nicer if we can just omit `@importFrom`, and make `@exportFrom` imply `@importFrom`:

```r
#' @exportFrom htmlwidgets JS saveWidget
#' @exportFrom magrittr %>%
```

**Comment by hadley** [comment]
The challenge is also making it work when someone has already imported the whole package.

**Comment by yihui** [comment]
Since you have the package name in `@exportFrom package fun`, in theory you can check if there is already an entry `import(package)` in NAMESPACE, and write out `importFrom(package, fun)` only if necessary.

**Comment by hadley** [comment]
That would be hard because the `@import` directive might come after the `@exportFrom`.

**Comment by yihui** [comment]
In that case, you can probably clean up the NAMESPACE entries in the final step (before you write them out), e.g. if you see both `import(package)` and `importFrom(package, foo)`, then drop the latter. I guess you don't write to NAMESPACE incrementally, do you?

**Comment by hadley** [comment]
No, but I don't have any representation of the namespace apart from a vector of strings, and it would be a lot of work to change that.

**Comment by hadley** [comment]
With @wch I talked about a tag called `@importFromAndExport` but that seemed a bit long.

**Comment by yihui** [comment]
Perhaps `@exportImportFrom`.

**Comment by hadley** [comment]
`@exportImportFrom` doesn't feel quite right to me because the order is wrong - you first import, then export the function, not the other way around.  How about this:

```R
#' @importAndExport
magrittr::`%>%`
```

?

**Comment by yihui** [comment]
That is okay, but I was also looking for a way to re-export multiple objects from a package on a single line, and that is why I was using `From`, like

```r
#' @exportImportFrom htmlwidgets JS saveWidget
```

It reads "export (those) imported from (package_name)". Or how about `@reexportFrom`?

```r
#' @reexportFrom htmlwidgets JS saveWidget
```

**Comment by hadley** [comment]
`@exportImportedFrom` would be ok. I don't like the double e in `reexport` - I think it makes it hard to parse (as a reader).

But I do like explicitly documenting the objects you're importing/exporting - it does take a bit more space but it's nice and explicit. And it also allows you to either `@export` or `@importAndExport`, depending on whether you've already imported it or not.

**Comment by hadley** [comment]
I vaguely remember there's also some other technical reason that implementing `@exportImportedFrom` is harder.

**Comment by hadley** [comment]
Hmmm, but maybe I can make both forms work

**Comment by hadley** [comment]
Ok, it's slowly coming back to - currently it's easy to add extra fields to an object with the `object_defaults()` function. But that means you need a special "type" of object to document. In this case, I think it's fairly natural to document a call to `::`. Making `@exportImportedFrom` work is harder because it gets parsed much later, and modifying the description etc at that point is much more complicated.

So in principle I like this idea - it would be nice to have a terser way of requesting an re-export, but I don't have time to get it working for this release. I'll leave this issue open, and then tackle it again when I next work on roxygen2.

**Comment by yihui** [comment]
Okay, take your time. It is not an urgent issue for me. The `#` issue was much more important, and I'm happy it has been fixed.

I don't like the double e, either. The tag name `@exportImportedFrom` sounds good.
